pkg/analysis/optionalfields/analyzer.go

```diff
@@ -183,10 +212,56 @@ func (a *analyzer) checkFieldPropertiesWithoutOmitEmpty(pass *analysis.Pass, fie
+		a.checkFieldPropertiesWithOmitEmptyRequired(pass, field, fieldName, jsonTags, underlying, false, hasValidZeroValue, completeValidation, isPointer, isStruct)
+	}
+}
+
+func (a *analyzer) checkFieldPropertiesWithOmitZeroRequired(pass *analysis.Pass, field *ast.Field, fieldName string, jsonTags extractjsontags.FieldTagInfo, hasOmitZero, isPointer, isStruct, hasValidZeroValue bool) {
+	if !isStruct || hasValidZeroValue {
+		return
+	}
+
+	a.handleFieldShouldHaveOmitZero(pass, field, fieldName, hasOmitZero, jsonTags)
+	a.handleFieldShouldNotBePointer(pass, field, fieldName, isPointer, "field %s is optional and does not have a valid zero value. The field does not need to be a pointer.")
+}
+
+func (a *analyzer) checkFieldPropertiesWithoutOmitZero(pass *analysis.Pass, field *ast.Field, fieldName string, underlying ast.Expr, jsonTags extractjsontags.FieldTagInfo, hasValidZeroValue, isPointer, isStruct, hasOmitEmpty, completeValidation bool) {
+	if !isStruct {
+		// Handle omitzero only for struct fields.
+		return
+	}
+
+	switch {
+	case hasValidZeroValue && !completeValidation:
+		// The field is not omitzero, and the zero value is valid, the field should be a pointer with omitempty tag.
+		a.handleIncompleteFieldValidation(pass, field, fieldName, isPointer, underlying)
+		fallthrough // Since it's a valid zero value, we should still enforce the pointer.
+	case hasValidZeroValue:
+		// Optional structs without omitzero should always be pointers and have omitempty, else they won't actually be omitted.
+		a.handleFieldShouldBePointer(pass, field, fieldName, isPointer, underlying)
+		a.handleFieldShouldHaveOmitEmpty(pass, field, fieldName, hasOmitEmpty, jsonTags)
+	case !hasValidZeroValue:
+		// The zero value would not be accepted, so the struct field needs to have omitzero.
+		// Force the omitzero policy to suggest a fix. We can only get to this function when the omitzero policy is configured to Ignore.
+		// Since we absolutely have to add the omitzero tag, we can report it as a suggestion.
+		reportShouldAddOmitZero(pass, field, OptionalFieldsOmitZeroPolicySuggestFix, fieldName, "field %s is optional and does not allow the zero value. It must have the omitzero tag.", jsonTags)
+
+		// Once it has the omitzero tag, it need not be a pointer in some cases.
+		// Now handle it as if it had the omitzero already.
+		// We already handle the omitzero tag above, so force the `hasOmitZero` to true.
+		a.checkFieldPropertiesWithOmitZeroRequired(pass, field, fieldName, jsonTags, true, isPointer, isStruct, hasValidZeroValue)
 	}
 }
 
+func (a *analyzer) checkFieldPropertiesWithOmitZeroForbidPolicy(pass *analysis.Pass, field *ast.Field, fieldName string, isStruct, hasOmitZero bool, jsonTags extractjsontags.FieldTagInfo) {
+	if !isStruct || !hasOmitZero {
+		// Handle omitzero only for struct field having omitZero tag.
+		return
+	}
+
+	reportShouldRemoveOmitZero(pass, field, fieldName, jsonTags)
```

Comment:
Does something else make sure it's pointer and omitempty? I guess the omitempty policy will handle it at this point, so we fall back to the existing behaviour, and this is therefore fine